src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -1063,77 +1342,150 @@ void ShenandoahFreeSet::find_regions_with_alloc_capacity(size_t &young_cset_regi
+          if (idx < mutator_leftmost) {
+            mutator_leftmost = idx;
+          }
+          if (idx > mutator_rightmost) {
+            mutator_rightmost = idx;
+          }
+          if (ac == region_size_bytes) {
+            if (idx < mutator_leftmost_empty) {
+              mutator_leftmost_empty = idx;
+            }
+            if (idx > mutator_rightmost_empty) {
+              mutator_rightmost_empty = idx;
+            }
+          }
+          mutator_regions++;
+          mutator_used += (region_size_bytes - ac);
+        }
       }
     }
   }
+  log_debug(gc)("  At end of prep_to_rebuild, mutator_leftmost: " SIZE_FORMAT
+                ", mutator_rightmost: " SIZE_FORMAT
+                ", mutator_leftmost_empty: " SIZE_FORMAT
+                ", mutator_rightmost_empty: " SIZE_FORMAT
+                ", mutator_regions: " SIZE_FORMAT
+                ", mutator_used: " SIZE_FORMAT,
+                mutator_leftmost, mutator_rightmost, mutator_leftmost_empty, mutator_rightmost_empty,
+                mutator_regions, mutator_used);
+
+  log_debug(gc)("  old_collector_leftmost: " SIZE_FORMAT
+                ", old_collector_rightmost: " SIZE_FORMAT
+                ", old_collector_leftmost_empty: " SIZE_FORMAT
+                ", old_collector_rightmost_empty: " SIZE_FORMAT
+                ", old_collector_regions: " SIZE_FORMAT
+                ", old_collector_used: " SIZE_FORMAT,
+                old_collector_leftmost, old_collector_rightmost, old_collector_leftmost_empty, old_collector_rightmost_empty,
+                old_collector_regions, old_collector_used);
+
+  _partitions.establish_mutator_intervals(mutator_leftmost, mutator_rightmost, mutator_leftmost_empty, mutator_rightmost_empty,
+                                          mutator_regions, mutator_used);
+  _partitions.establish_old_collector_intervals(old_collector_leftmost, old_collector_rightmost, old_collector_leftmost_empty,
+                                                old_collector_rightmost_empty, old_collector_regions, old_collector_used);
+  log_debug(gc)("  After find_regions_with_alloc_capacity(), Mutator range [" SSIZE_FORMAT ", " SSIZE_FORMAT "],"
+                "  Old Collector range [" SSIZE_FORMAT ", " SSIZE_FORMAT "]",
+                _partitions.leftmost(ShenandoahFreeSetPartitionId::Mutator),
+                _partitions.rightmost(ShenandoahFreeSetPartitionId::Mutator),
+                _partitions.leftmost(ShenandoahFreeSetPartitionId::OldCollector),
+                _partitions.rightmost(ShenandoahFreeSetPartitionId::OldCollector));
 }
 
-// Move no more than cset_regions from the existing Collector and OldCollector free sets to the Mutator free set.
-// This is called from outside the heap lock.
-void ShenandoahFreeSet::move_collector_sets_to_mutator(size_t max_xfer_regions) {
+void ShenandoahFreeSet::move_regions_from_collector_to_mutator(size_t max_xfer_regions) {
```

Comment:
This method has 3 isomorphic loops. I wonder if it might read better to extract a separate worker method for the loop and invoke it thrice from this method with appropriate parms. Also I'd advise using a local variable to countdown the regions to be transferred rather than modifying the method's input parm.

Reply:
I've factored out the loop bodies into two function calls.  Good suggestions.   Thanks.